### DISCUSSION TITLE
To collect a page range from the end of the document

### DISCUSSION THREAD

**Discussion by johnmapeson** [pdfcpu/pdfcpu#919] [Q&A] [answered]
Is it possible to collect a page range from the end of the document? That is, for example, from the 4th to the 2nd pages from the end, or simply the last 3 pages? In PDFtk, for example, it can be achieved using the letter `r`. For a 100-page document, the command `pdftk input-file.pdf cat 2-4 6-r6 r4-r2 output output-file.pdf` will extract the pages 2-4, 6-95, and 97-99.

I tried

- `pdfcpu collect -p -4--2 input3.pdf output.pdf`
- `pdfcpu collect -p l-4-l-2 input.pdf output.pdf`
- `pdfcpu collect -p l-4-2 input.pdf output.pdf`

But it doesn't seem to work.

**Comment by hhrutter** [comment]
See `pdfcpu selectedpages` on the CLI, this should work.
May be a bug - I need to look into it.
Thanks for reporting 💚

**Comment by johnmapeson** [comment]
Hello, Horst! I checked the man page carefully, but still cannot figure out how to select from the 4th to the 2nd pages from the end.

**Comment by hhrutter** [comment]
ok I'll check!

**Comment by hhrutter** [comment]
Have you tried `4-l-2` ?
This should select pages starting with 2 until the 2d page from the end (last - 2).

**Comment by johnmapeson** [comment]
> This should select pages starting with 2 until the 2d page from the end

If tested with a 100-page document, `pdfcpu collect -p 4-l-2 input.pdf output.pdf` removes pages 1-3 and 99-100 for me. The table below shows how to keep pages 2-4, 6-95 and 97-99 (and remove other pages) using other PDF processors. What I'm trying to figure out is how (and whether) it is possible using Pdfcpu.

| pdf processor             | example command                                                        |
|--------------|---------------------------------------------------------|
| pdftk        | `pdftk input.pdf cat 2-4 6-r6 r4-r2 output output.pdf`  |
| qpdf         | `qpdf input.pdf --pages . 2-4,6-r6,r4-r2 -- output.pdf` |
| coherent pdf | `cpdf input.pdf 2-4,6-~6,~4-~2 -o output.pdf`           |
| mupdf        | `mutool convert -o output.pdf input.pdf 2-4,6--6,-4--2` |
| pdfcpu       | ?                                                       |

**Comment by hhrutter** [comment, marked as answer]
These are alle equivalent and produce the same output:

`pdfcpu collect -p 2-,n5,nl-4,nl in.pdf out.pdf`
`pdfcpu collect -p 1-,n1,n5,nl-4,nl in.pdf out.pdf`

and using `!` for `not`:

`pdfcpu collect -p '2-,!5,!l-4,!l' in.pdf out.pdf`
`pdfcpu collect -p '1-,n1,n5,nl-4,nl' in.pdf out.pdf`

`!` on the command line needs to be quoted.

**Comment by johnmapeson** [comment]
Many thanks, Horst! This works, though syntax is not very intuitive :-)

**Comment by hhrutter** [comment]
Well, you take your pick!

Please consider becoming a [pdfcpu sponsor](https://github.com/sponsors/hhrutter) 💚
